pkg/cluster/cluster.go

```diff
@@ -205,6 +211,65 @@ func (p *Peer) Join(peerType PeerType, initialMetadata PeerMetadata) error {
 		Metadata:     initialMetadata,
 	})
 
+	if p.refreshInterval != 0 {
+		go p.periodicallyRefresh()
+	}
+
+	return nil
+}
+
+func (p *Peer) periodicallyRefresh() {
+	tick := time.NewTicker(p.refreshInterval)
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-p.stopc:
+			return
+		case <-tick.C:
+			if err := p.Refresh(); err != nil {
+				level.Error(p.logger).Log("msg", "Refreshing memberlist", "err", err)
+			}
+		}
+	}
+}
+
+// Refresh renews membership cluster, this will refresh DNS names and join newly added members
+func (p *Peer) Refresh() error {
+	p.mlistMtx.Lock()
+	defer p.mlistMtx.Unlock()
+
+	if p.mlist == nil {
+		return nil
+	}
+
+	resolvedPeers, err := resolvePeers(context.Background(), p.knownPeers, p.advertiseAddr, *net.DefaultResolver, false)
+	if err != nil {
+		return errors.Wrapf(err, "refresh cluster could not resolve peers: %v", resolvedPeers)
+	}
+
+	currMembers := p.mlist.Members()
+	for _, peer := range resolvedPeers {
+		var isPeerFound bool
+
+		for _, mem := range currMembers {
+			if mem.Address() == peer {
+				isPeerFound = true
+				break
+			}
+		}
+
+		if !isPeerFound {
+			curr, err := p.mlist.Join([]string{peer})
```

Comment:
why not looping through resolvedPeers and currMembers to find not connected addresses and just do `p.mlist.Join(notConnected)`? Might be easier to grasp. What do you think?

Reply:
Yeah, make sense.